reactfire/database/index.tsx

```diff
@@ -1,30 +1,43 @@
 import { database } from 'firebase/app';
 import { list, object, QueryChange, listVal } from 'rxfire/database';
-import {
-  ReactFireOptions,
-  useObservable,
-  checkIdField,
-  checkStartWithValue
-} from '..';
+import { ReactFireOptions, useObservable, checkIdField, checkStartWithValue } from '..';
 
 import { Observable } from 'rxjs';
 import { map } from 'rxjs/operators';
 
+const QUERY_UNIQUE_IDS = '_reactFireDatabaseQueryUniqueIds';
+
+// Since we're side-effect free, we need to ensure our observableId cache is global
+const cachedUniqueIds: Map<string, database.Query> = globalThis[QUERY_UNIQUE_IDS] || new Map();
+
+if (!globalThis[QUERY_UNIQUE_IDS]) {
+  globalThis[QUERY_UNIQUE_IDS] = cachedUniqueIds;
+}
+
+function getUnqiueIdForDatabaseQuery(query: database.Query) {
+  for (const [cachedUniqueId, cachedQuery] of cachedUniqueIds.entries()) {
+    if (cachedQuery.isEqual(query)) {
+      return cachedUniqueId;
+    }
+  }
+  let uniqueId: string;
+  do {
+    uniqueId = Math.random()
+      .toString(36)
+      .split('.')[1];
+  } while (cachedUniqueIds.has(uniqueId));
+  cachedUniqueIds.set(uniqueId, query);
```

Comment:
Couldn't you just use an array for `cachedUniqueIds ` and use the array index as the unique ID?

Reply:
Fair! I already had this code kicking around, so I reused it. Could just be as simple as an array thanks.

Reply:
Thanks again for helping get this sorted. Addressed.